Add UnreadyAddresses to Endpoints

smarterclayton opened the pull request:
In many cases clients may wish to view unready addresses for endpoints
in order to do set membership prior to a pod being ready. For instance,
a pod that uses the service endpoints to connect to other pods under
the same service, but does not want to signal ready before it has
contacted at least a minimal number of other pods.

This is backwards compatible with old servers and clients. There is
an additional cost in size of endpoints before services ramp up, which
will add minor CPU and memory use for services that have a significant
number of pods which have not become ready.

This will enable addition of other DNS service discovery endpoints which
could return all endpoints for a service.  The active set will not be
changed by this commit, nor will DNS be altered.

smarterclayton committed:
Add NotReadyAddresses to Endpoints

In many cases clients may wish to view not ready addresses for endpoints
in order to do set membership prior to a pod being ready. For instance,
a pod that uses the service endpoints to connect to other pods under
the same service, but does not want to signal ready before it has
contacted at least a minimal number of other pods.

This is backwards compatible with old servers and clients. There is
an additional cost in size of endpoints before services ramp up, which
will add minor CPU and memory use for services that have a significant
number of pods which have not become ready.

smarterclayton commented:
@thockin re: unready endpoints (allow clients to see their set prior to signally ready) and @kubernetes/kube-api

bgrant0607 commented on pkg/api/v1/types.go:
```diff
@@ -1649,6 +1649,8 @@ type Endpoints struct {
 	// IP addresses which offer the related ports.
 	Addresses []EndpointAddress `json:"addresses,omitempty"`
+	// IP addresses which offer the related ports but have not yet passed readiness checks
+	UnreadyAddresses []EndpointAddress `json:"unreadyAddresses,omitempty"`
```
Nice way to address backward compatibility without redundant info or a new virtual resource. I like it. We can consider refactoring the info in the v2 API -- we at least know we'll need to shard endpoints into multiple resources (in a stable fashion, such as by using subset and IP prefix).

smarterclayton replied:
Yeah, my primary concern was not upsetting the core path but still enabling the access.

karlkfi commented:
I understand the desire to include the NotReady endpoints in the list, and I assume you did it this way for reverse compatibility...

But honestly, I'd rather have a whole new API that uses the new semantics of "all known service targets", rather than suffer with `Endpoints.Subset.UnreadyAddresses` until the heat death of the universe (when we're allowed to remove deprecated APIs).

I was actually just brainstorming something like that to try to untangle the "full join" permutation mess that is `EndpointSubset`, but it's not quite ready to share yet.

karlkfi commented:
FWIW "NotReadyAddresses" would be consistent with the way kubectl prints Node readiness. 
`componentstatuses` uses Healthy/Unhealthy, but I consider that a bad example (for other reasons). I prefer the Condition/NotCondition semantics.

smarterclayton commented:
Yeah I wasn't a fan of Unready.  It's {Condition Ready False} in the API, NotReady in kubectl.  I don't think any of the conjunctions of ReadyFalse work well as a field name.

smarterclayton commented:
If anyone has anything better than NotReadyAddresses let me know.

smarterclayton commented:
v2 would be a new API and remove it, so anyone on the newer version wouldn't have to suffer the name. Introducing a new api has other implications so we've tended to evolve existing APIs by adding fields that fit into the existing scheme rather than create lots of new API objects (which also require documentation and increase user / client confusion).  I wouldn't assume v2 is so far away that additions to fields causes problems.

karlkfi commented:
I think "Not" is semantically used in kubectl to indicate both "Unknown" and "False". So as long as `Addresses` only includes "True", it makes sense to use "Not" for the inverse set.

karlkfi commented:
Does having a `Addresses` and `NotReadyAddresses` here imply that we can't add tracking "liveness" of endpoints?

`Endpoints` is implicitly a status-like model (as opposed to a spec), because the endpoint permutations and resulting packed structure are dynamically generated based on a cross between `PodSpec` and `ServiceSpec`. But there's no place to put a set of Conditions for each endpoint, because there's no denormalized `Endpoint` object to add it to.

And if Addresses means "ready addresses", then there's no way to add any more conditions, like liveness, in a reverse compatible way.

smarterclayton commented:
I think the question would be whether all the conditions of an endpoint are relevant.  Ready is a superset of Liveness, and today endpoints are only either Ready or not.  A not live endpoint is also not ready.

karlkfi commented:
If you know you only want to know liveness, and readiness, then you could probably make a structure that tells all three, similar to how you have but with an added NotAliveAddresses, then maybe dedupe NotReady ones and put them in NotAlive if that's the case.

However, you still wouldn't be able to have the extensibility that a conditions array has to add new conditions later.

smarterclayton commented:
I don't think the liveness check is intended to be displayed here.
Liveness is a formal constraint on the pod to ensure that it is always
responding.  Since the kubelet imposes the constraint that a pod that fails
liveness is restarted, and thus is marked as not-ready, I don't think we
intended liveness to be exposed into this model in that way.  Instead,
liveness is an input that determines whether something is Ready or NotReady.

If we think that there should be additional conditions on endpoints, I
would probably state those are imposed on the service (and thus be a
feature we add to a service to gather additional data, and then write that
back to endpoints).  Those sound more like "endpoint conditions" that are
imposed by the endpoints controller or a service health check controller,
and are then applied to endpoints.  In that model, we would still have a
separation between Ready and NotReady endpoints, but how endpoints show up
in either category depends on the admin.  We could certainly have a set of
conditions on each address (and it's possible that conditions on an
endpoint might be per port, not per address).




On Wed, Sep 9, 2015 at 11:40 PM, Karl Isenberg <notifications@github.com>
wrote:

> If you know you only want to know liveness, and readiness, then you could
> probably make a structure that tells all three, similar to how you have but
> with an added NotAliveAddresses, then maybe dedupe NotReady ones and put
> them in NotAlive if that's the case.
>
> However, you still wouldn't be able to have the extensibility that a
> conditions array has to add new conditions later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778#issuecomment-139107496>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
Looks much better as "notReady" and "NotReady"

smarterclayton commented:
Note I'm not trying to constrain what the V2 API model for endpoints might look like, just trying to focus this philosophically to something that concretely unblocks a use case in v1.

karlkfi commented:
> Note I'm not trying to constrain what the V2 API model for endpoints might look like, just trying to focus this philosophically to something that concretely unblocks a use case in v1.

Sure. I'm just kinda brainstorming here.

I think I'm coming around to `Addresses` and `NotReadyAddresses` being an acceptable compromise to gain the functionality. There should probably be some kind of disclaimer comment on the API tho, like `NotReadyAddresses was named as a compromise to preserve reverse compatibility and should not be used as an example for other APIs`.

> I don't think the liveness check is intended to be displayed here.

You're right that Pod liveness is not needed here. That wasn't what I was implying.

When I was writing the Component Registration PR, I was using component liveness to mean that the service was alive, and also responding to connection requests (required to verify liveness remotely), but not ready. The primary benefit of this is to distinguish between unreachable (NotAlive) and reachable but non-functional (NotReady).  Alive/NotReady means something different for remote components than it does for Pods, because Pod liveness is tested locally, which doesn't require network reachability or even a network listener (e.g. cmd-exec that checks a pid is running). When liveness is judged by a remote actor, liveness means that the networking layer is working and that the endpoint is reachable, and that the component is running and accepting connections. It's possible this should have a third label, rather than liveness, but liveness already optionally uses http/tcp. So renaming wouldn't be reverse compatible.

Anyway, remote Alive/NotReady could be useful for monitoring and notification, because NotReady is not a critical as NotAlive.  An Alive/NotReady endpoint might just be waiting for a dependency to be Ready and not require any intervention or alerting. It also makes root cause analysis easier to have more information about the reachability, to rule out networking issues.

That said, I think I'm ok with only having readiness of endpoints for the time being. I think switching to a structure that allows a list of conditions is desirable, for the above reasons, but I can't think of any way to do it that doesn't break reverse compatibility.

lavalamp commented:
I'm in favor of @smarterclayton's api change here, because it's easy, compatible and it gives useful information.

I also agree that if we want to add a third piece of information, incrementing the API version and making a list of conditions per endpoint is the way to go. And maybe making endpoint (singular) into an actual object. But that's a different topic.

smarterclayton commented:
Ok, now this is ready (e2e failure was a race to see endpoints be created)

smarterclayton commented:
@bgrant0607 did your previous ok constitute API approval?

derekwaynecarr commented on pkg/api/endpoints/util.go:
```diff
@@ -64,12 +65,37 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
 				}
 				portsToAddrs[*epp].Insert(allAddrs[ak])
 			}
+			for k := range subsets[i].NotReadyAddresses {
```
it feels like there is a lot of code duplication here with what is in subsets[i].Addresses, would be good if we could somehow make something common.

smarterclayton replied:
Ok

smarterclayton replied:
Removed duplication, also added comment that NotReady trumps Ready if we
see the same address twice.

On Fri, Sep 11, 2015 at 2:18 PM, Derek Carr <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39300442>
> :
>
> > @@ -64,12 +65,37 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
> >  				}
> >  				portsToAddrs[*epp].Insert(allAddrs[ak])
> >  			}
> > +			for k := range subsets[i].NotReadyAddresses {
>
> it feels like there is a lot of code duplication here with what is in
> subsets[i].Addresses, would be good if we could somehow make something
> common.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39300442>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

karlkfi commented on pkg/api/v1/types.go:
```diff
@@ -1649,6 +1649,8 @@ type Endpoints struct {
 type EndpointSubset struct {
 	// IP addresses which offer the related ports.
 	Addresses []EndpointAddress `json:"addresses,omitempty"`
+	// IP addresses which offer the related ports but have not yet passed readiness checks
```
Technically they could have stopped being ready (aka died) too, right? It's not just bootstrapping nodes. Can we clarify?

karlkfi commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -32,8 +32,8 @@ import (
 
-func addPods(store cache.Store, namespace string, nPods int, nPorts int) {
-	for i := 0; i < nPods; i++ {
+func addPods(store cache.Store, namespace string, nPods int, nPorts int, nUnready int) {
```
nNotReady for consistency?

karlkfi commented on pkg/controller/endpoint/endpoints_controller_test.go:
```diff
@@ -321,6 +324,81 @@ func TestSyncEndpointsItemsEmptySelectorSelectsAll(t *testing.T) {
 	endpointsHandler.ValidateRequest(t, testapi.Default.ResourcePath("endpoints", ns, "foo"), "PUT", &data)
 }
 
+func TestSyncEndpointsItemsEmptySelectorSelectsAllUnready(t *testing.T) {
```
NotReady

smarterclayton commented:
ok to test

smarterclayton commented:
ok to test

bgrant0607 commented on pkg/api/v1/types.go:
```diff
@@ -1649,6 +1649,10 @@ type Endpoints struct {
 	// IP addresses which offer the related ports.
 	Addresses []EndpointAddress `json:"addresses,omitempty"`
+	// IP addresses which offer the related ports but are not currently marked as ready
+	// because they have not yet finished starting, or have recently failed a liveness
```
readiness probe, not liveness check

smarterclayton replied:
Failing a liveness check will result in ready being cleared as well.  I'll add both.

bgrant0607 commented:
Yes, API is approved.

Sorry, I don't have time to read the full thread, but readiness is what most clients care about, and generally subsumes liveness. 

@ArtfulCoder PTAL

ArtfulCoder commented on api/swagger-spec/v1.json:
```diff
@@ -11228,6 +11228,13 @@
       },
       "description": "IP addresses which offer the related ports."
```
Are addresses considered ready if in the "addresses" array ?
Or are "addresses" the collection of both ready and notReady addresses ?

Is is possible for an address to ready if it is ready on all ports ?

Can the description answer these questions explicitly ?

smarterclayton replied:
Updated with more description.

On Tue, Sep 15, 2015 at 4:44 PM, Abhi Shah <notifications@github.com> wrote:

> In api/swagger-spec/v1.json
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39562448>
> :
>
> > @@ -11228,6 +11228,13 @@
> >        },
> >        "description": "IP addresses which offer the related ports."
>
> Are addresses considered ready if in the "addresses" array ?
> Or are "addresses" the collection of both ready and notReady addresses ?
>
> Is is possible for an address to ready if it is ready on all ports ?
>
> Can the description answer these questions explicitly ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39562448>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

thockin commented on pkg/api/types.go:
```diff
@@ -1305,8 +1305,9 @@ type Endpoints struct {
-	Addresses []EndpointAddress
-	Ports     []EndpointPort
+	Addresses         []EndpointAddress
+	NotReadyAddresses []EndpointAddress
```
Is this a one-of situation?  Is it valid to have Addresses and NotReadyAddresses at the same time in a single subset?

Comment pls.

smarterclayton replied:
Not in the same subset.  We split the subset

On Sep 15, 2015, at 8:36 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/types.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39583170>:

> @@ -1305,8 +1305,9 @@ type Endpoints struct {
>  //     a: [ 10.10.1.1:8675, 10.10.2.2:8675 ],
>  //     b: [ 10.10.1.1:309, 10.10.2.2:309 ]
>  type EndpointSubset struct {
> -	Addresses []EndpointAddress
> -	Ports     []EndpointPort
> +	Addresses         []EndpointAddress
> +	NotReadyAddresses []EndpointAddress

Is this a one-of situation? Is it valid to have Addresses and
NotReadyAddresses at the same time in a single subset?

Comment pls.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39583170>.

ArtfulCoder replied:
@smarterclayton Your response about splitting the subset does not seem to be in line with your code change. per your code-changes, it does seem that both arrays (Addresses and NotReadyAddresses) can be in the same subset, but one endpointAddress can only be in one or the other arrays.

ArtfulCoder replied:
you seem to split it in the controller code changes,but not in the repacksubset code

smarterclayton replied:
Repack subset *should* guarantee that an address doesn't show up in ready /
not ready in the same subset.  I'll update the doc, but I'm not sure what
code is indicating that we can't repack the subset to be unique.

On Wed, Sep 16, 2015 at 9:54 AM, Abhi Shah <notifications@github.com> wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39631558>
> :
>
> > @@ -1305,8 +1305,9 @@ type Endpoints struct {
> >  //     a: [ 10.10.1.1:8675, 10.10.2.2:8675 ],
> >  //     b: [ 10.10.1.1:309, 10.10.2.2:309 ]
> >  type EndpointSubset struct {
> > -	Addresses []EndpointAddress
> > -	Ports     []EndpointPort
> > +	Addresses         []EndpointAddress
> > +	NotReadyAddresses []EndpointAddress
>
> you seem to split it in the controller code changes,but not in the
> repacksubset code
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39631558>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

thockin commented on pkg/api/v1/types.go:
```diff
@@ -1631,7 +1631,10 @@ type Endpoints struct {
 
-	// The set of all endpoints is the union of all subsets.
+	// The set of all endpoints is the union of all subsets. Addresses are placed into
+	// subsets according to the IPs they share. If a single address has multiple ports,
```
What does it mean for a port to be ready?  AFAIU readiness is a property of the pod, not of a pod-port?

smarterclayton replied:
Multiple Containers

On Sep 15, 2015, at 8:37 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/v1/types.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39583236>:

> @@ -1631,7 +1631,10 @@ type Endpoints struct {
>  	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
>  	ObjectMeta `json:"metadata,omitempty"`
>
> -	// The set of all endpoints is the union of all subsets.
> +	// The set of all endpoints is the union of all subsets. Addresses are placed into
> +	// subsets according to the IPs they share. If a single address has multiple ports,

What does it mean for a port to be ready? AFAIU readiness is a property of
the pod, not of a pod-port?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39583236>.

thockin commented on pkg/api/v1/types.go:
```diff
@@ -1647,8 +1650,13 @@ type Endpoints struct {
+	// IP addresses which offer the related ports that are marked as ready. These endpoints
+	// should be considered safe for load balancers and clients to utilize.
 	Addresses []EndpointAddress `json:"addresses,omitempty"`
+	// IP addresses which offer the related ports but are not currently marked as ready
```
Comment needed: are Addresses and NotReadyAddresses mutually exclusive?

smarterclayton replied:
Yes, will add comment

On Sep 15, 2015, at 8:39 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/v1/types.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39583306>:

>  	Addresses []EndpointAddress `json:"addresses,omitempty"`
> +	// IP addresses which offer the related ports but are not currently marked as ready

Comment needed: are Addresses and NotReadyAddresses mutually exclusive?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39583306>.

thockin commented on pkg/api/endpoints/util.go:
```diff
@@ -34,42 +34,24 @@ import (
 	for i := range subsets {
 		for j := range subsets[i].Ports {
-			epp := &subsets[i].Ports[j]
+			epp := subsets[i].Ports[j]
```
why did you un-pointer it?

smarterclayton replied:
It was not used in a pointer context and the copy by value should be safe

thockin replied:
Safe and wasteful?  It's a minor point, and guidance for Go code varies...

On Tue, Sep 15, 2015 at 6:00 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39584404>
> :
>
> >  	allAddrs := map[addressKey]*api.EndpointAddress{}
> >  	portsToAddrs := map[api.EndpointPort]addressSet{}
> >  	for i := range subsets {
> >  		for j := range subsets[i].Ports {
> > -			epp := &subsets[i].Ports[j]
> > +			epp := subsets[i].Ports[j]
>
> It was not used in a pointer context and the copy by value should be safe
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39584404>.
>

smarterclayton replied:
The compiler *should* optimize away the allocation because it's a copy into
the map.  That said...

On Sep 15, 2015, at 9:08 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/endpoints/util.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39584956>:

>  	allAddrs := map[addressKey]*api.EndpointAddress{}
>  	portsToAddrs := map[api.EndpointPort]addressSet{}
>  	for i := range subsets {
>  		for j := range subsets[i].Ports {
> -			epp := &subsets[i].Ports[j]
> +			epp := subsets[i].Ports[j]

Safe and wasteful? It's a minor point, and guidance for Go code varies...
… <#>
On Tue, Sep 15, 2015 at 6:00 PM, Clayton Coleman <notifications@github.com>
wrote: In pkg/api/endpoints/util.go <
https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39584404> :
> allAddrs := map[addressKey]*api.EndpointAddress{} > portsToAddrs :=
map[api.EndpointPort]addressSet{} > for i := range subsets { > for j :=
range subsets[i].Ports { > - epp := &subsets[i].Ports[j] > + epp :=
subsets[i].Ports[j] It was not used in a pointer context and the copy by
value should be safe — Reply to this email directly or view it on GitHub <
https://github.com/kubernetes/kubernetes/pull/13778/files#r39584404>.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39584956>.

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
 
-type addressSet map[*api.EndpointAddress]struct{}
+// The sets of hosts must be de-duped, using IP+UID as the key.
+type addressKey struct {
```
I think we should just get rid of addressKey and use the endpointAddress as the key itself.
We are inconsistent in the code here. we use addressKey in the map, but not during hash generation of the address set.
We should be consistent.

smarterclayton replied:
It's unique ports here, but when we divvy by subsets we allocate by unique
host ready status.  But this we originally Tims code and I may be missing a
case that could be simplfied

On Sep 15, 2015, at 10:28 PM, Abhi Shah <notifications@github.com> wrote:

In pkg/api/endpoints/util.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39588390>:

>  	}
>
>  	// Finally, sort it.
>  	return SortSubsets(final)
>  }
>
> -type addressSet map[*api.EndpointAddress]struct{}
> +// The sets of hosts must be de-duped, using IP+UID as the key.
> +type addressKey struct {

I think we should just get rid of addressKey and use the endpointAddress as
the key itself.
We are inconsistent in the code here. we use addressKey in the map, but not
during hash generation of the address set.
We should be consistent.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39588390>.

smarterclayton replied:
So this is a question for Tim - Tim, why do we want to preserve duplicate IPs that come from different target refs?  That's what addressKey provides

ArtfulCoder replied:
i think it probably covers the case where docker subnet is the same on different nodes.

thockin replied:
Mesos.  Their IPs end up being the node's IP, but the pods are different.

On Wed, Sep 16, 2015 at 12:12 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39672451>
> :
>
> >  	}
> >
> >  	// Finally, sort it.
> >  	return SortSubsets(final)
> >  }
> >
> > -type addressSet map[*api.EndpointAddress]struct{}
> > +// The sets of hosts must be de-duped, using IP+UID as the key.
> > +type addressKey struct {
>
> So this is a question for Tim - Tim, why do we want to preserve duplicate
> IPs that come from different target refs? That's what addressKey provides
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39672451>.
>

karlkfi replied:
What @thockin said. On Mesos, Pod address is not a unique identifier (at least until IP per task is supported). address + port should be unique tho. There were a bunch of e2e tests we fixed recently that made this assumption too.

smarterclayton replied:
Ok, added very explicit comment

On Sep 16, 2015, at 11:06 PM, Karl Isenberg <notifications@github.com>
wrote:

In pkg/api/endpoints/util.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39708189>:

>  	}
>
>  	// Finally, sort it.
>  	return SortSubsets(final)
>  }
>
> -type addressSet map[*api.EndpointAddress]struct{}
> +// The sets of hosts must be de-duped, using IP+UID as the key.
> +type addressKey struct {

What @thockin <https://github.com/thockin> said. On Mesos, Pod address is
not a unique identifier (at least until IP per task is supported). address
+ port should be unique tho. There were a bunch of e2e tests we fixed
recently that made this assumption too.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39708189>.

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -34,42 +34,24 @@ import (
-					portsToAddrs[*epp] = addressSet{}
-				}
-				portsToAddrs[*epp].Insert(allAddrs[ak])
+				joinAddress(&subsets[i].Addresses[k], epp, true, allAddrs, portsToAddrs)
```
can the function joinAddress be renamed to something that better reflects what it is doing ?

smarterclayton replied:
how about mapAddressByPorts

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -34,42 +34,24 @@ import (
 	allAddrs := map[addressKey]*api.EndpointAddress{}
 	portsToAddrs := map[api.EndpointPort]addressSet{}
 	for i := range subsets {
 		for j := range subsets[i].Ports {
```
if you are going with a copy for epp, can you just say j, epp in the statement .

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+		var addrs, notReadyAddrs []api.EndpointAddress
+		for k, ready := range addrSets[key] {
```
can you rename k to address
all the single-letter variables in this function and the joinAddress function need to be renamed.
it is very difficult to follow what is going on...

smarterclayton replied:
I feel like I'm taking Tim's code review on the chin...

On Wed, Sep 16, 2015 at 5:34 PM, Abhi Shah <notifications@github.com> wrote:

> In pkg/api/endpoints/util.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39688300>
> :
>
> > @@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
> >  	// Next, build the N-to-M association the API wants.
> >  	final := []api.EndpointSubset{}
> >  	for key, ports := range addrSetsToPorts {
> > -		addrs := []api.EndpointAddress{}
> > -		for k := range addrSets[key] {
> > -			addrs = append(addrs, *k)
> > +		var addrs, notReadyAddrs []api.EndpointAddress
> > +		for k, ready := range addrSets[key] {
>
> can you rename k to address
> all the single-letter variables in this function and the joinAddress
> function need to be renamed.
> it is very difficult to follow what is going on...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39688300>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

ArtfulCoder commented on pkg/api/v1/types_swagger_doc_generated.go:
```diff
@@ -294,7 +295,7 @@ func (EndpointSubset) SwaggerDoc() map[string]string {
 	"":         "Endpoints is a collection of endpoints that implement the actual service. Example:\n  Name: \"mysvc\",\n  Subsets: [\n    {\n      Addresses: [{\"ip\": \"10.10.1.1\"}, {\"ip\": \"10.10.2.2\"}],\n      Ports: [{\"name\": \"a\", \"port\": 8675}, {\"name\": \"b\", \"port\": 309}]\n    },\n    {\n      Addresses: [{\"ip\": \"10.10.3.3\"}],\n      Ports: [{\"name\": \"a\", \"port\": 93}, ...
 	"metadata": "Standard object's metadata. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata",
-	"subsets":  "The set of all endpoints is the union of all subsets. Sets of addresses and ports that comprise a service.",
+	"subsets":  "The set of all endpoints is the union of all subsets. Addresses are placed into subsets according to the IPs they share. A single address with multiple ports, some of which are ready and some of which are not (because they come from different containers) will result in the address being displayed in different subsets for the different ports. No address will appear in both Addresses  ...
```
my understanding from your code is that, an address is either ready or not. We don't split the readiness of an address by port.

smarterclayton replied:
An address+port is ready or not.  An address that appears in one subset
will not appear in any other subset that has the same ports as the subset
the address is in.  An address can show up in two different subsets via
different ports, and have different readiness depending on whether that
particular port is ready.

On Wed, Sep 16, 2015 at 5:39 PM, Abhi Shah <notifications@github.com> wrote:

> In pkg/api/v1/types_swagger_doc_generated.go
> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39688882>
> :
>
> > @@ -294,7 +295,7 @@ func (EndpointSubset) SwaggerDoc() map[string]string {
> >  var map_Endpoints = map[string]string{
> >  	"":         "Endpoints is a collection of endpoints that implement the actual service. Example:\n  Name: \"mysvc\",\n  Subsets: [\n    {\n      Addresses: [{\"ip\": \"10.10.1.1\"}, {\"ip\": \"10.10.2.2\"}],\n      Ports: [{\"name\": \"a\", \"port\": 8675}, {\"name\": \"b\", \"port\": 309}]\n    },\n    {\n      Addresses: [{\"ip\": \"10.10.3.3\"}],\n      Ports: [{\"name\": \"a\", \"port\": 93}, {\"name\": \"b\", \"port\": 76}]\n    },\n ]",
> >  	"metadata": "Standard object's metadata. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata",
> > -	"subsets":  "The set of all endpoints is the union of all subsets. Sets of addresses and ports that comprise a service.",
> > +	"subsets":  "The set of all endpoints is the union of all subsets. Addresses are placed into subsets according to the IPs they share. A single address with multiple ports, some of which are ready and some of which are not (because they come from different containers) will result in the address being displayed in different subsets for the different ports. No address will appear in both Addresses and NotReadyAddresses in the same subset. Sets of addresses and ports that comprise a service.",
>
> my understanding from your code is that, an address is either ready or
> not. We don't split the readiness of an address by port.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39688882>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton replied:
Corrected the last vestiges of Tim's programming style

On Wed, Sep 16, 2015 at 5:42 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> An address+port is ready or not.  An address that appears in one subset
> will not appear in any other subset that has the same ports as the subset
> the address is in.  An address can show up in two different subsets via
> different ports, and have different readiness depending on whether that
> particular port is ready.
>
> On Wed, Sep 16, 2015 at 5:39 PM, Abhi Shah <notifications@github.com>
> wrote:
>
>> In pkg/api/v1/types_swagger_doc_generated.go
>> <https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39688882>
>> :
>>
>> > @@ -294,7 +295,7 @@ func (EndpointSubset) SwaggerDoc() map[string]string {
>> >  var map_Endpoints = map[string]string{
>> >  	"":         "Endpoints is a collection of endpoints that implement the actual service. Example:\n  Name: \"mysvc\",\n  Subsets: [\n    {\n      Addresses: [{\"ip\": \"10.10.1.1\"}, {\"ip\": \"10.10.2.2\"}],\n      Ports: [{\"name\": \"a\", \"port\": 8675}, {\"name\": \"b\", \"port\": 309}]\n    },\n    {\n      Addresses: [{\"ip\": \"10.10.3.3\"}],\n      Ports: [{\"name\": \"a\", \"port\": 93}, {\"name\": \"b\", \"port\": 76}]\n    },\n ]",
>> >  	"metadata": "Standard object's metadata. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata",
>> > -	"subsets":  "The set of all endpoints is the union of all subsets. Sets of addresses and ports that comprise a service.",
>> > +	"subsets":  "The set of all endpoints is the union of all subsets. Addresses are placed into subsets according to the IPs they share. A single address with multiple ports, some of which are ready and some of which are not (because they come from different containers) will result in the address being displayed in different subsets for the different ports. No address will appear in both Addresses and NotReadyAddresses in the same subset. Sets of addresses and ports that comprise a service.",
>>
>> my understanding from your code is that, an address is either ready or
>> not. We don't split the readiness of an address by port.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/13778/files#r39688882>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

thockin replied:
Readiness is a per-container attribute, not per-pod -- I missed that before, which is what lead to my question.  Got it now.

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -34,70 +34,93 @@ import (
 // form for things like comparison.  The result is a newly allocated slice.
 func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
```
// RepackSubsets takes a slice of EndpointSubset objects, expands it to the full
// representation, and then repacks that into the canonical layout.  This
// ensures that code which operates on these objects can rely on the common
// form for things like comparison.  The result is a newly allocated slice.
func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
	// First map each unique port definition to the sets of hosts that
	// offer it.
	allAddrs := map[addressKey]*api.EndpointAddress{}
	portToAddrReadyMap := map[api.EndpointPort]addressReadyMap{}
	for i := range subsets {
		for j := range subsets[i].Ports {
			epp := subsets[i].Ports[j]
			for k := range subsets[i].Addresses {
				joinAddress(&subsets[i].Addresses[k], epp, true, allAddrs, portToAddrReadyMap)
			}
			for k := range subsets[i].NotReadyAddresses {
				joinAddress(&subsets[i].NotReadyAddresses[k], epp, false, allAddrs, portToAddrReadyMap)
			}
		}
	}

	// Next, map the sets of hosts to the sets of ports they offer.
	// Go does not allow maps or slices as keys to maps, so we have
	// to synthesize an artificial key and do a sort of 2-part
	// associative entity.
	type keyString string
	keyToAddrReadyMap := map[keyString]addressReadyMap{}
	addrReadyMapKeyToPorts := map[keyString][]api.EndpointPort{}
	for epp, addrReadyMap := range portToAddrReadyMap {
		addrReadyMapKey := keyString(hashAddresses(addrReadyMap))
		keyToAddrReadyMap[addrReadyMapKey] = addrReadyMap
		addrReadyMapKeyToPorts[addrReadyMapKey] = append(addrReadyMapKeyToPorts[addrReadyMapKey], epp)
	}

	// Next, build the N-to-M association the API wants.
	final := []api.EndpointSubset{}
	for addrReadyMapKey, ports := range addrReadyMapKeyToPorts {
		var addrs, notReadyAddrs []api.EndpointAddress
		for address, addrReady := range keyToAddrReadyMap[addrReadyMapKey] {
			if addrReady {
				addrs = append(addrs, *address)
			} else {
				notReadyAddrs = append(notReadyAddrs, *address)
			}
		}
		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
	}

	// Finally, sort it.
	return SortSubsets(final)
}

ArtfulCoder replied:
I have re-written the function with better variable names.
can you use this ?

ArtfulCoder commented on pkg/api/endpoints/util.go:
```diff
@@ -34,70 +34,93 @@ import (
+	// if we have not yet recorded this port for this address, or if the previous
+	// state was ready, write the current ready state. not ready always trumps
+	// ready.
+	if wasReady, found := portsToAddrs[port][existingAddress]; !found || wasReady {
```
I think we should use addressKey as key, instead of existingAddress
Two existingAddress structs can be the same logically, but not be the same by go's perspective.

smarterclayton replied:
isn't that the point of allAddrs, to reduce that?  We want access to the full target ref later, so we still need addrMap to be populated.

ArtfulCoder replied:
ah. yes. the code is really difficult to follow along with.

smarterclayton replied:
Yes. Tim just sees deeper than the rest of us :)

On Sep 16, 2015, at 7:01 PM, Abhi Shah <notifications@github.com> wrote:

In pkg/api/endpoints/util.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39695860>:

> +	if existingAddress == nil {
> +		// Make a copy so we don't write to the
> +		// input args of this function.
> +		existingAddress = &api.EndpointAddress{}
> +		*existingAddress = *addr
> +		allAddrs[key] = existingAddress
> +	}
> +
> +	// Remember that this port maps to this address.
> +	if _, found := portsToAddrs[port]; !found {
> +		portsToAddrs[port] = addressSet{}
> +	}
> +	// if we have not yet recorded this port for this address, or if the previous
> +	// state was ready, write the current ready state. not ready always trumps
> +	// ready.
> +	if wasReady, found := portsToAddrs[port][existingAddress]; !found || wasReady {

ah. yes. the code is really difficult to follow along with.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39695860>.

ArtfulCoder commented:
changes are good. will wait for the variable renaming.
lgtm otherwise.

smarterclayton commented:
Updated with new names, I added a few more comments.

On Wed, Sep 16, 2015 at 7:02 PM, Abhi Shah <notifications@github.com> wrote:

> changes are good. will wait for the variable renaming.
> lgtm otherwise.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13778#issuecomment-140920055>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

thockin commented on pkg/api/endpoints/util.go:
```diff
@@ -34,70 +34,96 @@ import (
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+	for key, ports := range addrReadyMapKeyToPorts {
+		var addrs, notReadyAddrs []api.EndpointAddress
```
s/addrs/readyAddrs

ArtfulCoder commented:
LGTM
will need rebase

smarterclayton commented:
Simple rebase, readding lgtm

karlkfi commented:
Can't tell on my phone, but can you make sure the e2e tests cover sufficient cases now that there are more permutations? I'd like to run the conformance tests, but I know they don't cover load balancing. So we need to be extra sure this does what we think it does.

smarterclayton commented:
What are you concerned about specifically?  The unit tests cover all the new behavior.  I don't know what other e2e test to add beyond verifying that an unready endpoint shows up.  Some suggestions of tests to write would help.

karlkfi commented:
Removed lgtm to delay merging a bit. I want to take another look at e2e when I get to my computer and run the conformance tests on our mesos CI to make sure there weren't any undocumented assumptions that got lost. I'll do it today.

smarterclayton commented:
Merge queue beat you.  Does the MesosCI run automatically or is it manual?

karlkfi commented:
Manual on PRs (for now), auto/continuously on Master. 

Looks like it's run since it was merged and there's no new e2e failures. But I don't see any e2e tests that actually exercise NotReadyEndpoints. So that doesn't really give me a lot of confidence.

I would love an e2e test with a pod container whose Readiness can be toggled (start 500 erroring on the readiness probe endpoint) via http request while remaining Alive. Then you could spin up a replication controller of 3 pods, make one NotReady, see that the endpoint are adjusted accordingly, make it Ready again, and see endpoints updated again.

smarterclayton commented:
e2e/kubectl.go verifies that a not ready address shows up, then disappears.

smarterclayton commented:
Agree it could be better - I'll spawn a follow up issue.

smarterclayton commented:
No wait, that was the endpoints controller unit test.  Follow up issue will mention adding an e2e.
